lib/rules/require-sorted-object-keys.js

```diff
@@ -0,0 +1,118 @@
+/**
+ * Requires sorted object keys.
+ *
+ * Types: `Boolean` or `String`
+ *
+ * Values:
+ *  - `true` (alias to `asc`)
+ *  - `"asc"`: requires sorting in ascending order
+ *  - `"desc"`: requires sorting in descending order
+ *
+ * #### Example
+ *
+ * ```js
+ * "requireSortedObjectKeys": "asc"
+ * ```
+ *
+ * ##### Valid
+ *
+ * ```js
+ * var x = {
+ *  x: 'foo',
+ *  y: 'bar'
+ * }
+ * ```
+ *
+ * ##### Invalid
+ *
+ * ```js
+ * var x = {
+ *  y: 'foo',
+ *  x: 'bar'
+ * }
+ * ```
+ */
+
+function asc(a, b) {
+    if (/[_]/.test(a)) {
+        return -1;
+    }
+
+    if (a < b) {
+        return -1;
+    }
+
+    if (a > b) {
+        return 1;
+    }
+
+    return 0;
+}
+
+function desc(a, b) {
+    if (/[_]/.test(a)) {
+        return 1;
+    }
+
+    if (/[_]/.test(b)) {
+        return -1;
+    }
+
+    if (a > b) {
+        return -1;
+    }
+
+    if (a < b) {
+        return 1;
+    }
+
+    return 0;
+}
+
+var assert = require('assert');
+
+module.exports = function() {};
+
+module.exports.prototype = {
+
+    configure: function(options) {
+        assert(
+            options === true || options === 'asc' || options === 'desc',
+            this.getOptionName() + ' option requires a true value or should be removed'
+        );
+
+        this.sort = options === true ? 'asc' : options;
+    },
+
+    getOptionName: function() {
+        return 'requireSortedObjectKeys';
+    },
+
+    check: function(file, errors) {
+        var sort = this.sort;
+
+        file.iterateNodesByType('ObjectExpression', function(node) {
+            var keys = node.properties.map(function(property) {
+                return property.key.name;
+            });
+
+            var sorted = keys.slice(0).sort(sort === 'asc' ? asc : desc);
```

Comment:
The best optimisation I see for now (but in my opinion in a different phase), would be to compare each property key name with the next one and immediately break out of the loop if a non-expected key is found instead of sorting all keys. I've thought about it before submitting the PR but I wasn't particularly enjoying the implementation I had.